DIPs/1NNN-ADR.md

```diff
@@ -0,0 +1,479 @@
+
+We may also provide user-friendly aliases for `_d_interpolated_string` and `_d_interpolated_format_spec`, but the color of the bikeshed is not important to this DIP author.
+
+### Justifications
+
+The library wrapper may seem superfluous, however it serves three key roles:
+
+1. It divorces the compiler entirely from details of generated format strings.
+2. It provides necessary error checking capabilities.
+3. It provides an additional API for user functions to introspect the string, building on D's existing compile-time capabilities.
+
+The previous version of this DIP proposed a simple string literal instead of the result of `_d_interpolated_string` as the first argument passed. This is problematic because it introduces potentially subtle errors in likely usage.
+
+#### Wrong-use in unrelated function
+
+Consider a function in D today:
+
+```
+/++
+	Creates a window with the given title, width, and height. If
+	width or height are set to zero (the default), your window will
+	be automatically sized.
++/
+Window createWindow(string title, int width = 0, int height = 0);
+```
+
+A user has some code that calls it like so:
+
+```
+import std.conv;
+auto window = createWindow("Process debugger " ~ to!string(pid));
+```
+
+A new version of the compiler is released and the user, eager to try D's new string interpolation feature, rewrites that code as would be common in many other languages:
+
+```
+// look at this clean new code :D
+auto window = createWindow(i"Process debugger $pid");
+```
+
+It compiles without error. In bliss, the happy user thinks: "d rox". Then... the program is run and the window width is extraordinary. But the documentation says it will be automatically sized by default. The confused user wonders: why did this change? They run the program again and get another random width. Is the library incompatible with the new dmd version?
+
+After some time, they notice the window title has changed to "Process debugger %" before being cut off by the taskbar width. "This new dmd version must be super buggy," they think, "it corrupted my string!"
+
+Exasperated, they go online to ask the D gurus if anyone else has experienced this bizarre behavior. After an hour of wasted time, they finally mention having experimented with D's new string interpolation feature earlier in the day.
+
+The helpful people in the D forum point them at the spec: "I know it looks like a string," the Master explains, "but it is actually lowered to an auto-expanding tuple whose first element is a string. That's causing your problem. I'd love to fix it in the library, but it is literally impossible to differentiate this case from legitimate usage :("
+
+"D is weird," the user replies, deflated and defeated, "I think I'm just going to use Python. Have you see its f-strings?"
```

Comment:
This is too much anecdotal here. A comparison and description with python would be useful here.

Reply:
oh that's just me having some fun referencing an IRC conversation, one of the users in there I was bouncing ideas off of kept referencing python so I threw it in there for my light-hearted amusement.

If you wanna know though the python f string is just like `f"Hello, {name}. You are {age}."` and it yields a plain old string, not that different than the typical dynamic language string. It does have the advantage of being simple to explain to people though and fits well for dynamic use cases which is why the frustrated user in the narrative went back to it after giving up on D's weirdness.

Of course, this whole section is a rebuttal to Walter's (now rejected) DIP; with the proposal in this document, such things do NOT occur.

Reply:
I agree with @12345swordy, the point can be made without the dialogue in the story. Just identify what would normally be expected, vs. what would happen.

Reply:
y'all are no fun :(

Reply:
Trust me I'm fun (come to dconf, you will see). But I also want to provide no excuses for rejection.

Reply:
I think we met in 2015! Though I didn't stick around after hours; after walking with most yall to the hotel I changed direction and visited a friend in Orem.

I'm considering going to the 2020 dconf... but I kinda hate flying, it is going the wrong way time zone wise, it means leaving the country... so idk, I guess i need to decide soonish though to put together a talk proposal just I'm not excited about it.

anyway i changed the paragraphs there in the recent commit

Reply:
I've been going since 2013 (and in 2014 the infamous "single page slide written last night", it was a work of art). I've enjoyed your talks.

The plane ride sucks, especially going that direction.